### PR TITLE
feat(lambda): add support for ephemeral storage

### LINKED ISSUE #19605: (lambda): add support for ephemeral storage setting
### Description

Add support for new larger ephemeral storage in lambdas. 

### Use Case

Lambdas often need more then 512MB to work with for temporary storage. In our case building files that are then moved to S3 but are built on the lambda first. Also simple tasks such as zipping/unzipping files before or after S3. 

### Proposed Solution

Add L2 property for ephemeralStorage in lambda constructor. 

### Other information

https://aws.amazon.com/blogs/aws/aws-lambda-now-supports-up-to-10-gb-ephemeral-storage/

### Acknowledge

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by robertd**
Depends on cfnspec bump PR https://github.com/aws/aws-cdk/pull/19553

CFN Docs: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-lambda-function-ephemeralstorage.html

Blog: https://aws.amazon.com/blogs/aws/aws-lambda-now-supports-up-to-10-gb-ephemeral-storage/

image

Closes #19605.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
	* [ ] Did you use `cdk-integ` to deploy the infrastructure and generate the snapshot (i.e. `cdk-integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-lambda/README.md
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-lambda/README.md:715]:
```suggestion
You can configure ephemeral storage on a function to control the amount of storage it gets for reading
or writing data, allowing you to use AWS Lambda for ETL jobs, ML inference, or other data-intensive workloads.
The ephemeral storage will be accessible in the functions' `/tmp` directory.
```

**Review by kaizen3031593 — CHANGES_REQUESTED** [review]
Hi @robertd! Thanks for being on top of this feature! Please add a test that shows that `ephemeralStorageSize` works and also one to show that an unresolved token does not throw an error.

File: packages/@aws-cdk/aws-lambda/lib/function.ts
Comment (robertd) [reply packages/@aws-cdk/aws-lambda/lib/function.ts:103]: Cool. TIL that `cdk.Size` is a thing :)

File: packages/@aws-cdk/aws-lambda/lib/function.ts
Comment (robertd) [reply packages/@aws-cdk/aws-lambda/lib/function.ts:1143]: Sounds good. I was trying to follow the existing coding style of the construct 👍

File: packages/@aws-cdk/aws-lambda/lib/function.ts
Comment (robertd) [reply packages/@aws-cdk/aws-lambda/lib/function.ts:777]: Agreed. I prefer this type of assignment but I wanted to follow the overall coding style of the construct. I’ll switch to this model.

**Comment by robertd** [comment]
@kaizen3031593 Do we need integ tests? I don’t see any for `memorySize` so I didn’t bother. Lmk

File: packages/@aws-cdk/aws-lambda/lib/function.ts
Comment (robertd) [reply packages/@aws-cdk/aws-lambda/lib/function.ts]: Removed.

File: packages/@aws-cdk/aws-lambda/lib/function.ts
Comment (robertd) [reply packages/@aws-cdk/aws-lambda/lib/function.ts:103]: Hi @kaizen3031593 I'm finally back at my computer. Comparison operators don't work with `Size` type... I need to figure out how to check for out-of-bound.

Comment (robertd) [reply packages/@aws-cdk/aws-lambda/lib/function.ts:103]: I'm looking at `Size` tests... will figure it out shortly. Please disregard my previous comment.

Comment (robertd) [reply packages/@aws-cdk/aws-lambda/lib/function.ts:103]: @kaizen3031593 Back to your original comment... should we move this to `Size` at all? Will that break developer experience given that `memorySize` is already a number?

**Comment by robertd** [comment]
> @kaizen3031593 Back to your original comment... should we move this to Size at all? Will that break developer experience given that memorySize is already a number?

> @kaizen3031593 Do we need integ tests? I don’t see any for memorySize so I didn’t bother. Lmk

@kaizen3031593 Any thoughts?

File: packages/@aws-cdk/aws-lambda/lib/function.ts
Comment (madeline-k) [reply packages/@aws-cdk/aws-lambda/lib/function.ts:103]: I think this should be a `cdk.Size`

File: packages/@aws-cdk/aws-lambda/README.md
Comment (madeline-k) [inline packages/@aws-cdk/aws-lambda/README.md:727]:
```suggestion

Read more about using this feature in [this AWS blog post](https://aws.amazon.com/blogs/aws/aws-lambda-now-supports-up-to-10-gb-ephemeral-storage/).
```

File: packages/@aws-cdk/aws-lambda/test/function.test.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-lambda/test/function.test.ts:2490]: Can you add 2 more unit tests? 
1. Showing that the property works, just the normal happy path.
2. Showing that you can use an unresolved token as the input for ephemeralStorageSize?

**Review by madeline-k — DISMISSED** [review]
Hey @robertd , @kaizen3031593 is on vacation this week. I'll be taking over the review here. I am good to approve after you switch to using `cdk.Size` and add a couple unit tests.

File: packages/@aws-cdk/aws-lambda/test/function.test.ts
Comment (robertd) [reply packages/@aws-cdk/aws-lambda/test/function.test.ts:2492]: @madeline-k  #1 is done... but can you elaborate more on why #2 is needed (learning moment for me). Also, if you could point me to an example test that would be awesome 😃 . TY.

Comment (madeline-k) [reply packages/@aws-cdk/aws-lambda/test/function.test.ts:2492]: Here is an example: https://github.com/aws/aws-cdk/blob/a81a0a45d72bf297c60ad01c7e5f84ea11b8cb61/packages/%40aws-cdk/aws-ssm/test/parameter.test.ts#L87

Comment (madeline-k) [reply packages/@aws-cdk/aws-lambda/test/function.test.ts:2492]: You can do this to create an unresolved token: 
```
cdk.Size.gibibytes(cdk.Lazy.number({ produce: () => 1 }));
```

Comment (madeline-k) [reply packages/@aws-cdk/aws-lambda/test/function.test.ts:2492]: I am not sure why exactly @kaizen3031593 requested it in this case. But in general, we test for this to make sure that validation doesn't prevent users from passing in unresolved tokens. Like CfnOutputs from other stacks.

Comment (robertd) [reply packages/@aws-cdk/aws-lambda/test/function.test.ts:2492]: TY!

Comment (robertd) [reply packages/@aws-cdk/aws-lambda/test/function.test.ts:2492]: @madeline-k I tried using example you posted but I keep getting this error:
```ts
test('ephemeral storage allows unresolved tokens', () => {
  const stack = new cdk.Stack();
  expect(() => {
    new lambda.Function(stack, 'MyLambda', {
      code: new lambda.InlineCode('foo'),
      handler: 'bar',
      runtime: lambda.Runtime.NODEJS_14_X,
      ephemeralStorageSize:  Size.gibibytes(Lazy.number({ produce: () => 1 })),
    });
  }).not.toThrow();
});
```
```bash
FAIL test/function.test.js (13.295 s)
@aws-cdk/aws-lambda:   ● ephemeral storage allows unresolved tokens
@aws-cdk/aws-lambda:     Unable to perform time unit conversion on un-resolved token -1.8881545897092687e+289.
@aws-cdk/aws-lambda:       191 |   if (fromUnit.inKibiBytes === toUnit.inKibiBytes) { return amount; }
@aws-cdk/aws-lambda:       192 |   if (Token.isUnresolved(amount)) {
@aws-cdk/aws-lambda:     > 193 |     throw new Error(`Unable to perform time unit conversion on un-resolved token ${amount}.`);
@aws-cdk/aws-lambda:           |           ^
@aws-cdk/aws-lambda:       194 |   }
@aws-cdk/aws-lambda:       195 |
@aws-cdk/aws-lambda:       196 |   const multiplier = fromUnit.inKibiBytes / toUnit.inKibiBytes;
@aws-cdk/aws-lambda:       at convert (../core/lib/size.ts:193:11)
@aws-cdk/aws-lambda:       at Size.toMebibytes (../core/lib/size.ts:112:12)
@aws-cdk/aws-lambda:       at new Function (lib/function.ts:754:38)
@aws-cdk/aws-lambda:       at Object.<anonymous> (test/function.test.ts:2526:3)
```
I also saw this test implementation, but that didn't work either: 
https://github.com/aws/aws-cdk/blob/f19896be08209f140fdf07189b9d1e4c9d601bc6/packages/%40aws-cdk/core/test/size.test.ts#L12

I guess I just am not really sure what I'm testing here tbh. :)

**Review by rix0rrr — DISMISSED** [review]

File: packages/@aws-cdk/aws-lambda/lib/function.ts
Comment (kaizen3031593) [reply packages/@aws-cdk/aws-lambda/lib/function.ts:753]: Was just going to comment this :). `Size.isUnresolved()` is introduced in [this](https://github.com/aws/aws-cdk/pull/19569) PR that is almost in. 

Btw I asked for testing for token support in this PR because it is a possible use case and a common bug we introduce in the CDK. Any time we are validating something we should think about the use case when a token is involved.

Comment (robertd) [reply packages/@aws-cdk/aws-lambda/lib/function.ts]: I merged in this change but now I'm getting this error:

>  The declaring class is introduced here
lib/function.ts(757,67): error TS2339: Property 'isUnresolved' does not exist on type 'Size'.

Comment (robertd) [reply packages/@aws-cdk/aws-lambda/lib/function.ts]: I just saw PR @kaizen3031593  mentioned... I'll wait until that gets merged in.

File: packages/@aws-cdk/aws-lambda/test/function.test.ts
Comment (madeline-k) [reply packages/@aws-cdk/aws-lambda/test/function.test.ts:2640]: With an unresolved token, you need to test with something in KibiBytes, since that is the target. @rix0rrr actually just made a change to make this error message much more helpful :)

Comment (madeline-k) [reply packages/@aws-cdk/aws-lambda/test/function.test.ts:2640]: Actually, I think the target should be MebiBytes in this case. So I'm a bit confused why the error you get is from a conversion to KibiBytes.

Comment (robertd) [reply packages/@aws-cdk/aws-lambda/test/function.test.ts:2640]: Yup... waiting for https://github.com/aws/aws-cdk/pull/19569 to get merged in.

**Comment by robertd** [comment]
@kaizen3031593 @madeline-k @rix0rrr PR ready for review. Thanks in advance.

**Review by madeline-k — APPROVED** [review]
Looks good to me. Thanks, @robertd!

**Comment by khoatrandata** [comment]
thanks for the work @robertd , @madeline-k ! Wonder any chance this can be released in a couple days?

**Comment by robertd** [comment]
@khoatrandata Most likely 2.20.0 will be released by the end of this week. CDK team is good at releasing at least once a week.

**Pull request merged**
